File: config/src/main/java/de/eschoenawa/lanchat/config/ConfigImpl.java
package de.eschoenawa.lanchat.config;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import de.eschoenawa.lanchat.config.persist.ConfigPersister;
import de.eschoenawa.lanchat.config.persist.SimpleFileConfigPersister;
import de.eschoenawa.lanchat.util.Log;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ConfigImpl implements Config {
    private static final String TAG = "Config";

    private final ConfigPersister persister;
    private final SettingsDefinition settingsDefinition;
    private final Map<String, Setting> settings = new HashMap<>();
    private Map<String, String> currentTransaction;

    public ConfigImpl(SettingsDefinition settingsDefinition) {
        this.settingsDefinition = settingsDefinition;
        loadSettingsFromDefinition();
        this.persister = new SimpleFileConfigPersister(getString(settingsDefinition.getConfigPathKey(), null));
        loadSettingsFromFile();
        this.settingsDefinition.validateSettings();
    }

    public ConfigImpl(SettingsDefinition settingsDefinition, ConfigPersister persister) {
        this.settingsDefinition = settingsDefinition;
        loadSettingsFromDefinition();
        this.persister = persister;
        loadSettingsFromFile();
        this.settingsDefinition.validateSettings();
    }

    private void loadSettingsFromDefinition() {
        for (Setting setting : settingsDefinition.getAllSettings()) {
            this.settings.put(setting.getKey(), setting);
        }
    }

    private void loadSettingsFromFile() {
        this.persister.load(new ConfigPersister.ConfigPersisterLoadCallback() {
            private boolean overwriteRequired = false;
            @Override
            public void onSettingLoaded(String key, String value) {
                Setting currentSetting = settings.get(key);
                if (currentSetting == null || !currentSetting.isModifiable()) {
                    Log.w(TAG, "Setting '" + key + "' is either not defined or not modifiable and therefore won't be loaded from file.");
                    overwriteRequired = true;
                } else {
                    currentSetting.setValue(value);
                    settings.put(key, currentSetting);
                }
            }

            @Override
            public void onPersistentSettingsNotYetCreated() {
                Log.d(TAG, "Config file doesn't exist yet. Creating new one now...");
                saveSettings();
                Log.d(TAG, "Config file created!");
            }

            @Override
            public void onLoadFailed(String reason) {
                Log.e(TAG, "Unable to read from config file (" + reason + ")! Using defaults.");
            }

            @Override
            public void onLoadDone() {
                if (overwriteRequired) {
                    Log.w(TAG, "Config file contains invalid or deprecated settings. Re-creating config file...");
                    saveSettings();
                    Log.d(TAG, "Config file re-created.");
                }
            }
        });
    }

    private void saveSettings() {
        Log.d(TAG, "Saving config file...");
        this.persister.save(generateModifiableSettingsMap());
        Log.d(TAG, "Config file saved!");
    }

    private Map<String, String> generateStringSettingsMap() {
        Map<String, String> result = new HashMap<>();
        for (String key : settings.keySet()) {
            Setting setting = settings.get(key);
            if (setting != null && setting.isModifiable()) {
                result.put(key, setting.getValue());
            }
        }
        return result;
    }

    @Override
    public synchronized void beginTransaction() {
        currentTransaction = new HashMap<>();
    }

    @Override
    public synchronized void abortTransaction() {
        currentTransaction = null;
    }

    @Override
    public synchronized void commitTransaction() {
        if (isTransactionActive()) {
            for (String key : currentTransaction.keySet()) {
                Setting setting = settings.get(key);
                if (setting != null && setting.isModifiable()) {
                    String newValue = currentTransaction.get(key);
                    Log.d(TAG, "Changing setting '" + key + "' to '" + newValue + "'.");
                    setting.setValue(newValue);
                    settings.put(key, setting);
                } else {
                    Log.w(TAG, "Ignoring changes to setting '" + key + "' (value set to '" + currentTransaction.get(key) + "')! Not registered: " + (setting == null));
                }
            }
            currentTransaction = null;
            saveSettings();
            settingsDefinition.validateSettings();
        } else {
            throw new IllegalStateException("No transaction currently active!");
        }
    }

    @Override
    public synchronized boolean isTransactionActive() {
        return currentTransaction != null;
    }

    @Override
    public synchronized boolean doesTransactionRequireRestart() {
        if (isTransactionActive()) {
            for (String key : currentTransaction.keySet()) {
                Setting setting = settings.get(key);
                if (setting != null && setting.isModifiable() && setting.isRestartRequired()) {
                    return true;
                }
            }
            return false;
        } else {
            throw new IllegalStateException("No transaction currently active!");
        }
    }

    @Override
    public List<Setting> getModifiableSettings() {
        List<Setting> result = new ArrayList<>();
        for (Setting setting : settings.values()) {
            if (setting.isModifiable()) {
                result.add(setting);
            }
        }
        return result;
    }

    @Override
    public String getString(String key, String def) {
        String result = null;
        if (isTransactionActive()) {
            result = currentTransaction.get(key);
        }
        if (result == null) {
            Setting found = settings.get(key);
            result = (found == null) ? def : found.getValue();
        }
        return result;
    }

    @Override
    public int getInt(String key, int def) {
        return Integer.parseInt(getString(key, String.valueOf(def)));
    }

    @Override
    public boolean getBoolean(String key, boolean def) {
        return Boolean.parseBoolean(getString(key, String.valueOf(def)));
    }

    @Override
    public double getDouble(String key, double def) {
        return Double.parseDouble(getString(key, String.valueOf(def)));
    }

    @Override
    public float getFloat(String key, float def) {
        return Float.parseFloat(getString(key, String.valueOf(def)));
    }

    @Override
    public long getLong(String key, long def) {
        return Long.parseLong(getString(key, String.valueOf(def)));
    }

    @Override
    public char getChar(String key, char def) {
        return getString(key, String.valueOf(def)).charAt(0);
    }

    @Override
    public String requireString(String key) {
        String result = getString(key, null);
        if (result == null) {
            throw new IllegalStateException("Value for key '" + key + "' is required but not stored in config!");
        }
        return result;
    }

    @Override
    public int requireInt(String key) {
        return Integer.parseInt(requireString(key));
    }

    @Override
    public boolean requireBoolean(String key) {
        return Boolean.parseBoolean(requireString(key));
    }

    @Override
    public double requireDouble(String key) {
        return Double.parseDouble(requireString(key));
    }

    @Override
    public float requireFloat(String key) {
        return Float.parseFloat(requireString(key));
    }

    @Override
    public long requireLong(String key) {
        return Long.parseLong(requireString(key));
    }

    @Override
    public char requireChar(String key) {
        return requireString(key).charAt(0);
    }

    @Override
    public synchronized void setString(String key, String value) {
        if (isTransactionActive()) {
            currentTransaction.put(key, value);
        } else {
            throw new IllegalStateException("No transaction currently active!");
        }
    }

    @Override
    public synchronized void setInt(String key, int value) {
        setString(key, String.valueOf(value));
    }

    @Override
    public synchronized void setBoolean(String key, boolean value) {
        setString(key, String.valueOf(value));
    }

    @Override
    public synchronized void setDouble(String key, double value) {
        setString(key, String.valueOf(value));
    }

    @Override
    public synchronized void setFloat(String key, float value) {
        setString(key, String.valueOf(value));
    }

    @Override
    public synchronized void setLong(String key, long value) {
        setString(key, String.valueOf(value));
    }

    @Override
    public synchronized void setChar(String key, char value) {
        setString(key, String.valueOf(value));
    }

    @Override
    public String toString() {
        Gson gson = new GsonBuilder().create();
        return gson.toJson(generateStringSettingsMap());
    }

    private Map<String, Setting> generateModifiableSettingsMap() {
        HashMap<String, Setting> result = new HashMap<>();
        for (String key : settings.keySet()) {
            Setting setting = settings.get(key);
            if (setting.isModifiable()) {
                result.put(key, setting);
            }
        }
        return result;
    }
}

File: config/src/main/java/de/eschoenawa/lanchat/config/ConfigImpl.kt
package de.eschoenawa.lanchat.config

import com.google.gson.GsonBuilder
import de.eschoenawa.lanchat.config.persist.ConfigPersister
import de.eschoenawa.lanchat.config.persist.ConfigPersister.ConfigPersisterLoadCallback
import de.eschoenawa.lanchat.config.persist.SimpleFileConfigPersister
import de.eschoenawa.lanchat.util.Log

class ConfigImpl : Config {
    companion object {
        private const val TAG = "Config"
    }

    private val persister: ConfigPersister
    private val settingsDefinition: SettingsDefinition
    private val settings: MutableMap<String, Config.Setting> = HashMap()
    private var currentTransaction: MutableMap<String, String?>? = null

    constructor(settingsDefinition: SettingsDefinition) {
        this.settingsDefinition = settingsDefinition
        loadSettingsFromDefinition()
        persister = SimpleFileConfigPersister(getString(settingsDefinition.configPathKey, null)!!)
        loadSettingsFromFile()
        this.settingsDefinition.validateSettings()
    }

    constructor(settingsDefinition: SettingsDefinition, persister: ConfigPersister) {
        this.settingsDefinition = settingsDefinition
        loadSettingsFromDefinition()
        this.persister = persister
        loadSettingsFromFile()
        this.settingsDefinition.validateSettings()
    }

    private fun loadSettingsFromDefinition() {
        for (setting in settingsDefinition.allSettings) {
            settings[setting.key] = setting
        }
    }

    private fun loadSettingsFromFile() {
        persister.load(object : ConfigPersisterLoadCallback {
            private var overwriteRequired = false
            override fun onSettingLoaded(key: String, value: String) {
                val currentSetting = settings[key]
                if (currentSetting == null || !currentSetting.isModifiable) {
                    Log.w(
                        TAG,
                        "Setting '$key' is either not defined or not modifiable and therefore won't be loaded from file."
                    )
                    overwriteRequired = true
                } else {
                    currentSetting.value = value
                    settings[key] = currentSetting
                }
            }

            override fun onPersistentSettingsNotYetCreated() {
                Log.d(TAG, "Config file doesn't exist yet. Creating new one now...")
                saveSettings()
                Log.d(TAG, "Config file created!")
            }

            override fun onLoadFailed(reason: String) {
                Log.e(TAG, "Unable to read from config file ($reason)! Using defaults.")
            }

            override fun onLoadDone() {
                if (overwriteRequired) {
                    Log.w(TAG, "Config file contains invalid or deprecated settings. Re-creating config file...")
                    saveSettings()
                    Log.d(TAG, "Config file re-created.")
                }
            }
        })
    }

    private fun saveSettings() {
        Log.d(TAG, "Saving config file...")
        persister.save(generateModifiableSettingsMap())
        Log.d(TAG, "Config file saved!")
    }

    private fun generateStringSettingsMap(): Map<String, String> {
        val result: MutableMap<String, String> = HashMap()
        for (key in settings.keys) {
            val setting = settings[key]
            if (setting != null && setting.isModifiable) {
                result[key] = setting.value
            }
        }
        return result
    }

    @Synchronized
    override fun beginTransaction() {
        currentTransaction = HashMap()
    }

    @Synchronized
    override fun abortTransaction() {
        currentTransaction = null
    }

    @Synchronized
    override fun commitTransaction() {
        currentTransaction?.let {transaction ->
            for (key in transaction.keys) {
                val setting = settings[key]
                if (setting != null && setting.isModifiable) {
                    val newValue = transaction[key]
                    Log.d(TAG, "Changing setting '$key' to '$newValue'.")
                    setting.value = newValue!!
                    settings[key] = setting
                } else {
                    Log.w(
                        TAG,
                        "Ignoring changes to setting '$key' (value set to '${transaction[key]}')! Not registered: ${setting == null}"
                    )
                }
            }
            currentTransaction = null
            saveSettings()
            settingsDefinition.validateSettings()
        } ?: throw IllegalStateException("No transaction currently active!")
    }

    @get:Synchronized
    override val isTransactionActive: Boolean
        get() = currentTransaction != null

    @Synchronized
    override fun doesTransactionRequireRestart(): Boolean {
        currentTransaction?.let { transaction ->
            for (key in transaction.keys) {
                val setting = settings[key]
                if (setting != null && setting.isModifiable && setting.isRestartRequired) {
                    return true
                }
            }
            return false
        } ?: throw IllegalStateException("No transaction currently active!")
    }

    override fun getModifiableSettings(): List<Config.Setting> {
        val result: MutableList<Config.Setting> = ArrayList()
        for (setting in settings.values) {
            if (setting.isModifiable) {
                result.add(setting)
            }
        }
        return result
    }

    override fun getString(key: String, def: String?): String? {
        var result: String? = null
        currentTransaction?.let { transaction ->
            result = transaction[key]
        } ?: run {
            val found = settings[key]
            result = found?.value ?: def
        }
        return result
    }

    override fun getInt(key: String, def: Int): Int {
        return getString(key, def.toString())!!.toInt()
    }

    override fun getBoolean(key: String, def: Boolean): Boolean {
        return getString(key, def.toString()).toBoolean()
    }

    override fun getDouble(key: String, def: Double): Double {
        return getString(key, def.toString())!!.toDouble()
    }

    override fun getFloat(key: String, def: Float): Float {
        return getString(key, def.toString())!!.toFloat()
    }

    override fun getLong(key: String, def: Long): Long {
        return getString(key, def.toString())!!.toLong()
    }

    override fun getChar(key: String, def: Char): Char {
        return getString(key, def.toString())!![0]
    }

    override fun requireString(key: String): String {
        return getString(key, null)
            ?: throw IllegalStateException("Value for key '$key' is required but not stored in config!")
    }

    override fun requireInt(key: String): Int {
        return requireString(key).toInt()
    }

    override fun requireBoolean(key: String): Boolean {
        return requireString(key).toBoolean()
    }

    override fun requireDouble(key: String): Double {
        return requireString(key).toDouble()
    }

    override fun requireFloat(key: String): Float {
        return requireString(key).toFloat()
    }

    override fun requireLong(key: String): Long {
        return requireString(key).toLong()
    }

    override fun requireChar(key: String): Char {
        return requireString(key)[0]
    }

    @Synchronized
    override fun setString(key: String, value: String?) {
        currentTransaction?.let { transaction ->
            transaction[key] = value
        } ?: throw IllegalStateException("No transaction currently active!")
    }

    @Synchronized
    override fun setInt(key: String, value: Int) {
        setString(key, value.toString())
    }

    @Synchronized
    override fun setBoolean(key: String, value: Boolean) {
        setString(key, value.toString())
    }

    @Synchronized
    override fun setDouble(key: String, value: Double) {
        setString(key, value.toString())
    }

    @Synchronized
    override fun setFloat(key: String, value: Float) {
        setString(key, value.toString())
    }

    @Synchronized
    override fun setLong(key: String, value: Long) {
        setString(key, value.toString())
    }

    @Synchronized
    override fun setChar(key: String, value: Char) {
        setString(key, value.toString())
    }

    override fun toString(): String {
        val gson = GsonBuilder().create()
        return gson.toJson(generateStringSettingsMap())
    }

    private fun generateModifiableSettingsMap(): Map<String, Config.Setting> {
        val result = HashMap<String, Config.Setting>()
        for (key in settings.keys) {
            val setting = settings[key]
            if (setting!!.isModifiable) {
                result[key] = setting
            }
        }
        return result
    }
}